execution/src/sce_ledger.rs

```diff
@@ -0,0 +1,332 @@
+use std::sync::{Arc, Mutex};
+
+use crate::ExecutionError;
+use crypto::hash::Hash;
+use models::ModelsError;
+use models::{address::AddressHashMap, hhasher::HHashMap, Address, Amount, AMOUNT_ZERO};
+use wasmer::Module;
+
+/// an entry in the SCE ledger
+#[derive(Debug, Clone, Default)]
+pub struct SCELedgerEntry {
+    pub balance: Amount,
+    pub opt_module: Option<Module>,
+    pub data: HHashMap<Hash, Vec<u8>>,
+}
+
+impl SCELedgerEntry {
+    /// applies an entry update to self
+    pub fn apply_entry_update(&mut self, update: &SCELedgerEntryUpdate) {
+        // balance
+        if let Some(new_balance) = update.update_balance {
+            self.balance = new_balance;
+        }
+
+        // module
+        if let Some(opt_module) = &update.update_opt_module {
+            self.opt_module = opt_module.clone();
+        }
+
+        // data
+        for (data_key, data_update) in update.update_data.iter() {
+            match data_update {
+                Some(new_data) => {
+                    self.data.insert(*data_key, new_data.clone());
+                }
+                None => {
+                    self.data.remove(data_key);
+                }
+            }
+        }
+    }
+}
+
+// optional updates to be applied to a ledger entry
+#[derive(Debug, Clone)]
+pub struct SCELedgerEntryUpdate {
+    pub update_balance: Option<Amount>,
+    pub update_opt_module: Option<Option<Module>>,
+    pub update_data: HHashMap<Hash, Option<Vec<u8>>>, // None for row deletion
+}
+
+impl SCELedgerEntryUpdate {
+    /// apply another SCELedgerEntryUpdate to self
+    pub fn apply_entry_update(&mut self, other: &SCELedgerEntryUpdate) {
+        // balance
+        if let Some(new_balance) = other.update_balance {
+            self.update_balance = Some(new_balance);
+        }
+
+        // module
+        if let Some(new_opt_module) = &other.update_opt_module {
+            self.update_opt_module = Some(new_opt_module.clone());
+        }
+
+        // data
+        self.update_data.extend(other.update_data.clone());
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum SCELedgerChange {
+    // delete an entry
+    Delete,
+
+    // sets an entry to an absolute value
+    Set(SCELedgerEntry),
+
+    // updates an entry
+    Update(SCELedgerEntryUpdate),
+}
+
+impl SCELedgerChange {
```

Comment:
I think almost all the logic in this module would be best done in a follow-up, and come with a detailed spec.

Reply:
it's basically done. I followed the same concepts as the consensus ledger, together with the specs of the VM (ledger change cancellability)